### ISSUE TITLE
How To Reference An Element in Shadow DOM?

### ISSUE THREAD

**Issue by auzmartist** [hybridsjs/hybrids#86]
Is there a way to reference an element in the shadow DOM? I am interested in using Hybrids to create WebGL components but I am struggling to see where I could get access to the canvas element in my template. 

For example:
```javascript
const GlContext = {
  width: '100%',
  height: '100%',
  render: ({width, height}) => html`
    <canvas class="gl-context" style="width: ${width}; height: ${height};">
    </canvas>
  `,
}
```
In the above component, is there a way to know when the canvas is rendered so that I can select it in the DOM and return the webgl2 context? I'd imagine that somewhere, after render, I could do:

```javascript
const canvas = host.shadowRoot.querySelector('canvas.gl-context')
const gl = canvas.getContext('webgl2')
```

The inverse approach to lifecycle hooks seems to run against the grain of what I'm trying to do. Any way to obtain the reference to the element in the template?

**Comment by smalluban** [comment]
Thanks for creating this issue - your problem inspired me to make some fixes and unlock easier way to create refs of the internal parts from the template. You can look at the referenced PR, open [updated render factory docs from the branch](https://github.com/hybridsjs/hybrids/blob/fix-render-refs/docs/built-in-factories/render.md#manual-update) and try it out on the [codesandbox.io](https://codesandbox.io/s/hybrids-simple-counter-w3n5m) live before code will land on npm (it uses own built version of the package from the branch).

> The inverse approach to lifecycle hooks seems to run against the grain of what I'm trying to do.

However, there are two interesting issues in the subject of referencing internal parts of the render or calling DOM after rendering. I recommend you to read them, as I try to explain there the idea behind the concepts of the library:

https://github.com/hybridsjs/hybrids/issues/78
https://github.com/hybridsjs/hybrids/issues/56

Let me know what do you think about solution from the PR (you can comment there) and if this solves your problem.

**Comment by auzmartist** [comment]
Yes, thanks for the references to previous related issues. I ran through those examples you provided there but was somewhat unsatisfied with the implications of the documented workarounds. I eventually settled on the workaround described in #78 but am happy to see you've exposed some new functionality. Thank you for such a delightful library.

I'll test out the solution in the PR in a bit and give some more feedback if necessary. At first glance, this looks perfect. 🎉

**Comment by smalluban** [comment]
Fixed in #87

**Comment by dobexx** [comment]
Hello,

We are currently migrating approx. 250 hybrid elements from version 5 to 8. A few questions have arisen in the process.

For referencing elements in the Shadow DOM we have used the following factory functions so far:

`function ref(selector) {
	 return ({ render }) => render().querySelector(selector);
}`

and:

`function ref(query) {
  return ({ render }) => {
	 if (typeof render === 'function') {
		const target = render();
		return target.querySelector(query);
	 }

	 return null;
  };
}`

Now, however, there is no longer the "render" function. What would be an alternative way of implementing this?

Many thanks
